src/lib/model.js

```diff
@@ -54,21 +54,50 @@ const processGroups = data => data.reduce(
   , []);
 const groupUsage = (dateFrom, dateTo) => query.usageByGroup(dateFrom, dateTo).then(processGroups);
 
+const hours = name => ({ name, data: Array.from({ length: 24 }, (_, n) => ({ count: 0, hour: n })) });
+const dayCounts = (from, to) => (
+  (start, end,
+   diff = Math.abs(Math.round(moment.duration(start.startOf('day').diff(end.endOf('day'))).as('days'))),
+   mod = diff % 7,
+   weeks = Math.floor(diff / 7),
+   startDay = start.day() || 7,
+   extraWeekEndDays = mod && [6, 7].filter(n => startDay <= n && (startDay + mod) > n).length,
+   we = (weeks * 2) + extraWeekEndDays
+  ) => [diff - we, we]
+)(moment(from, DATE_FORMAT), moment(to, DATE_FORMAT));
+const processHourlyUsage = (nod, data) => {
+  const total = nod[0] + nod[1];
+  const traces = [nod[0] && hours('weekday'), nod[1] && hours('weekend')];
+  const average = hours('average').data;
+  data.forEach(d => {
+    traces[d.weekend].data[d.hour].count = (d.count / nod[d.weekend]);
+    average[d.hour].count += d.count;
+  });
+  return [...traces, nod[0] && nod[1] && {
+    name: 'average', data: average.map(o => ({ ...o, count: o.count / total }))
+  }].filter(e => e);
+};
+const hourlyUsage = (dateFrom, dateTo, searchGroup) => query.hourlyUsage(dateFrom, dateTo, searchGroup)
+  .then(processHourlyUsage.bind(null, dayCounts(dateFrom, dateTo || moment().endOf('day'))));
+
```

Comment:
I don't understand this approach. - Why not just use `GROUP BY` in a SQL query?

This looks like re-implementing something that is provided to us by the database.

Reply:
Looks like I've misunderstood what this code does as you are using `GROUP BY` elsewhere.

What is the purpose of this code? It looks very complicated.

I think the use of IIFE in `hours` makes it a lot harder to understand.

I think we discussed some of this code in Slack at some point. Is it worth looking at that discussion?

Reply:
I could look at shifting more work on to the DB... we should really have the report use the read-only DB endpoint, then if we have HA it won't be interfering with the main service.
Let me do some query-splaining and see what the difference is...

Reply:
Having looked at moving some of the implementation into the DB query, I don't think we can reasonably rely on that as an approach. I think as we can't be certain of finding at least one result a day for each day in the intended range, we can't use the DB. So I'll stick with the current code, at least for now.

Reply:
Well that issue would be fairly easy to fix but I agree it would need to be done in the JS code. (Well you could probably do it in SQL with a join but I'm not sure that's actually better.)

I think the bigger problem is that it's not clear what this code is actually doing. Possibly I just needs some comments to say why you are doing what you are doing. But I really think that the IIFE should go.

Reply:
I've refactored the IIFE. Fixed